### PR TITLE
loki: backend: use single dataframe for logs

### PR DESCRIPTION
**Pull request by gabor**
https://github.com/grafana/grafana/pull/47153 changed the dataframe-format for Loki. this PR applies this change to backend-mode.

details of the long PR:
- all the `go` code: we generate the new dataframe there. we moved the id-field generation to the backend from the frontend. we need this change for being able to handle live-streaming (not part of this pull request). we use the `fnv` hash that is supposed to be fast. it generates a 4bytes-long hash, we use the `nanosecondTimestamp_4bytesHash` as the `ID`
- `backendResultTransformer.ts` ( + the test file): adjusted to handle the new format. it parses the labels-structures from string to javasrcript-object
- `datasource.ts`: adjusted to "handle" live-streaming requests when backend-mode is enabled. it will still run the live-streaming requests through the frontend code-path, but at least it will work when backend-mode is enabled
- `makeIdField.ts` ( + the test file): removed, `ID` is now generated on the backend

### CODE REVIEW

**Comment by leeoniya** [comment]
> we use the fnv hash that is supposed to be fast

some other maybe faster options:

https://news.ycombinator.com/item?id=10673868

https://github.com/cespare/xxhash
https://github.com/pierrec/xxHash

**Comment by gabor** [comment]
> > we use the fnv hash that is supposed to be fast
> 
> some other maybe faster options:
> 
> https://news.ycombinator.com/item?id=10673868
> 
> https://github.com/cespare/xxhash https://github.com/pierrec/xxHash

thanks for the links!

for now i wanted something that does need additional dependencies, but we can certainly improve here.

(well, what i'd love to have is to not-have the id-field at all, to be honest, i just don't know how to do that 😄 )

**Review by ivanahuckova — APPROVED** [review]
Nice LGTM! Left a question not related to this PR below. We can discuss it later.

File: pkg/tsdb/loki/frame.go [hunk 1]
```diff
@@ -33,48 +62,109 @@ func adjustFrame(frame *data.Frame, query *lokiQuery) *data.Frame {
...
+}
+
+func adjustLogsFrame(frame *data.Frame, query *lokiQuery) error {
+	// we check if the fields are of correct type and length
+	fields := frame.Fields
+	if len(fields) != 3 {
+		return fmt.Errorf("invalid fields in logs frame")
+	}
+
+	labelsField := fields[0]
```
[the remark this answers is not in the document]
Comment (gabor) [reply pkg/tsdb/loki/frame.go:87]: good point. there's some conflicting needs between what live-streaming wants in some implementations and what all-other-non-live-streaming-things do 😄 , i'll flip it in this one to be the normal  time-value-labels. (the problem is, how much code is needed to "find" the labels-field)

**Pull request merged**